### PR TITLE
[15.0][MIG] sale project service tracking copy tasks

### PR DESCRIPTION
**Pull request by EmilioPascual**
This module add the possibility to create task in an existing project from the sales order through setting the product with a project template with the desired task.

@moduon  MT-1028

@Shide, @rafaelbn please review 😊

### CODE REVIEW

**Review by rafaelbn — APPROVED** [review]
👍🏼

File: sale_project_service_tracking_copy_tasks/models/product_template.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+from odoo import fields, models
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    service_tracking = fields.Selection(
+        selection_add=[
+            ("copy_tasks_in_project", "Copy tasks into sales order's project"),
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/models/product_template.py]:
```suggestion
            ("copy_tasks_in_project", "Copy tasks into sale order's project"),
```

File: sale_project_service_tracking_copy_tasks/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.depends("order_line.product_id.service_tracking")
+    def _compute_visible_project(self):
+        """Users should be able to select a project_id on the order if
+        at least one order line has a product with
+        its service tracking configured as 'copy_tasks_in_project'"""
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/models/sale_order.py]: **nitpick:** Did you know there's [PEP 257](https://peps.python.org/pep-0257/#multi-line-docstrings) for docstrings?

This looks much better:

```suggestion
        """Let users select a project on matching orders.

        Affects orders where at least one line has a product with
        its service tracking configured as 'copy_tasks_in_project'.
        """
```

File: sale_project_service_tracking_copy_tasks/models/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        result = super()._compute_visible_project()
+        for order in self:
+            if any(
+                service_tracking == "copy_tasks_in_project"
+                for service_tracking in order.order_line.mapped(
+                    "product_id.service_tracking"
+                )
+            ):
+                order.visible_project = True
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/models/sale_order.py]: **suggestion:** This is simpler:

```suggestion
            order.visible_project = "copy_tasks_in_project" in order.order_line.mapped(
                "product_id.service_tracking"
            ):
```

File: sale_project_service_tracking_copy_tasks/models/sale_order.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        for order in self:
+            projects = order.mapped("order_line.product_id.project_template_id")
+            project_data = projects[0].copy_data(
+                default=dict(
+                    type_ids=projects.mapped("type_ids").ids,
+                    active=True,
+                    **order._get_order_project_data(),
+                )
+            )
+            new_project = pp_model.create(project_data)
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/models/sale_order.py]: **suggestion:** this is simpler

```suggestion
            new_project = projects[0].copy(
                dict(
                    type_ids=projects.mapped("type_ids").ids,
                    active=True,
                    **order._get_order_project_data(),
                )
            )
```

File: sale_project_service_tracking_copy_tasks/models/sale_order.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            analytic_account_id=self.analytic_account_id.id,
+            name=self.name,
+        )
+
+    @api.returns("project.project")
+    def create_order_project(self):
+        pp_model = self.env["project.project"]
+        created_projects = pp_model.browse()
+        for order in self:
+            projects = order.mapped("order_line.product_id.project_template_id")
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/models/sale_order.py]: **question:** shouldn't you filter the products by their service tracking?

File: sale_project_service_tracking_copy_tasks/models/sale_order_line.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    def _timesheet_service_generation(self):
+        """Adds task inheritance from project_templates
+
+        Creates the project in sale order if needed and copies tasks
+        from product project_template to order project"""
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/models/sale_order_line.py]: **nitpick:** again improving docstring formatting:

```suggestion
        from product project_template to order project.
        """
```

File: sale_project_service_tracking_copy_tasks/models/sale_order_line.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        so_model = self.env["sale.order"]
+        mt_note = self.env.ref("mail.mt_note")
+        # Check lines that needs to inherit project template tasks
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/models/sale_order_line.py]:
```suggestion
        # Check lines that need to inherit project template tasks
```

File: sale_project_service_tracking_copy_tasks/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2023 Moduon Team S.L.
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl-3.0)
+
+{
+    "name": "Service tracking: Copy tasks in project",
+    "summary": "Copy tasks into sale order's project",
+    "version": "15.0.1.0.0",
+    "author": "Moduon, Odoo Community Association (OCA)",
+    "maintainers": ["shide", "EmilioPascual"],
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/__manifest__.py]: **chore(blocking):** add rafaelbn

See internal contribution docs.

File: sale_project_service_tracking_copy_tasks/readme/DESCRIPTION.rst [hunk 9]
```diff
@@ -0,0 +1,3 @@
+This module add the possibility to create task in an existing project from the sales order through setting the product with a project template with the desired task.
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/readme/DESCRIPTION.rst]:
```suggestion
This module allows creating tasks in an existing project from the sales order through setting the product with a project template with the desired task.
```

File: sale_project_service_tracking_copy_tasks/readme/DESCRIPTION.rst [hunk 10]
```diff
@@ -0,0 +1,3 @@
...
+
+The use case is when a Odoo customer sell the same services with the same task constantly and the task should be linked with existing project. Odoo don't cover this use case as Odoo always create a new project.
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/readme/DESCRIPTION.rst]: **nitpick:** add this to a `CONTEXT.rst` file instead.

```suggestion
The use case is when an Odoo customer sells the same services with the same tasks constantly and the tasks should be linked with existing project. Standard Odoo would always create a new project.
```

File: sale_project_service_tracking_copy_tasks/views/product_template_views.xml [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+<!-- Copyright 2023 Moduon Team S.L.
+     License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl-3.0) -->
+<odoo>
+    <record
+        id="product_template_form_view_service_tracking_copy_tasks_in_project"
+        model="ir.ui.view"
+    >
+        <field
+            name="name"
+        >product.template.service.tracking.copy.tasks.in.project.inherit</field>
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/views/product_template_views.xml]: **nitpick:** you can set a human-readable name in this field:

```suggestion
        >Display project when product is copy_tasks_in_project</field>
```

File: sale_project_service_tracking_copy_tasks/views/product_template_views.xml [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        <field name="model">product.template</field>
+        <field
+            name="inherit_id"
+            ref="sale_project.product_template_form_view_invoice_policy_inherit_sale_project"
+        />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='project_template_id']" position="attributes">
+                <attribute
+                    name="attrs"
+                >{'invisible': [('service_tracking', 'not in', ['task_in_project', 'project_only', 'copy_tasks_in_project'])], 'required': [('service_tracking', 'in', ['copy_tasks_in_project'])]}</attribute>
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/views/product_template_views.xml]: **thought(non-blocking):** https://github.com/OCA/server-tools/tree/16.0/base_view_inheritance_extension could let you add this in a more resilient fashion.

File: sale_project_service_tracking_copy_tasks/views/sale_order_views.xml [hunk 13]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!-- Copyright 2023 Moduon Team S.L.
+     License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl-3.0) -->
+<odoo>
+    <record model="ir.ui.view" id="view_order_form_inherit_sale_project">
+        <field name="model">sale.order</field>
```
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/views/sale_order_views.xml]:
```suggestion
        <field name="model">Support for project with copied tasks</field>
```

File: sale_project_service_tracking_copy_tasks/__manifest__.py
Comment (yajo) [inline sale_project_service_tracking_copy_tasks/__manifest__.py:1]: **thought(non-blocking):** IMHO the module name is a bit too large. What about `sale_project_copy_tasks`?

**Review by yajo — CHANGES_REQUESTED** [review]
Looks really good.

A few suggestions, but only a few are blocking.

File: sale_project_service_tracking_copy_tasks/models/sale_order.py (hunk 5 above)
Comment (EmilioPascual) [reply sale_project_service_tracking_copy_tasks/models/sale_order.py]: Filtered in _timesheet_service_generation from sale_order_line before the method is called

File: sale_project_service_tracking_copy_tasks/__manifest__.py (hunk 8 above)
Comment (EmilioPascual) [reply sale_project_service_tracking_copy_tasks/__manifest__.py]: Added him to CONTRIBUTORS.rst

File: sale_project_service_tracking_copy_tasks/views/product_template_views.xml (hunk 11 above)
Comment (EmilioPascual) [reply sale_project_service_tracking_copy_tasks/views/product_template_views.xml]: Name of views is identical to xml id with dots: https://www.odoo.com/documentation/15.0/es/contributing/development/coding_guidelines.html#security-view-and-action
Maybe you think it was an action.
In any case, I havve changed xml id and name

File: sale_project_service_tracking_copy_tasks/views/sale_order_views.xml (hunk 13 above)
Comment (EmilioPascual) [reply sale_project_service_tracking_copy_tasks/views/sale_order_views.xml]: I think you assumed it was the name, not the model.
I have added missing name.

**Review by Shide — APPROVED** [review]
Code LGTM

File: sale_project_copy_tasks/__manifest__.py [hunk 14]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2023 Moduon Team S.L.
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl-3.0)
+
+{
+    "name": "Service tracking: Copy tasks in project",
+    "summary": "Copy tasks into sale order's project",
+    "version": "15.0.1.0.0",
+    "author": "Moduon, Odoo Community Association (OCA)",
+    "maintainers": ["shide", "EmilioPascual"],
```
Comment (yajo) [inline sale_project_copy_tasks/__manifest__.py:9]:
```suggestion
    "maintainers": ["rafaelbn", "shide", "EmilioPascual"],
```

**Review by yajo — CHANGES_REQUESTED** [review]
Just one missing detail to go.

**Review by yajo — APPROVED** [review]

**Pull request merged**